### ISSUE TITLE
Feature request: Support for Wayland's wlr-layer-shell protocol

### ISSUE THREAD

**Issue by Keating950** [rust-windowing/winit#2582]
I would find it really helpful if winit provided a friendly interface for creating [layer shells](https://wayland.app/protocols/wlr-layer-shell-unstable-v1) in Wayland, perhaps as an extension of WindowBuilder. So far, I haven't been able to find a solution for this in the Rust ecosystem other than by going to the relatively low-level [SCTK](https://github.com/Smithay/client-toolkit). I'm imagining this feature could be gated off in one of the traits in [winit::platform::unix](https://docs.rs/winit/latest/winit/platform/unix/index.html) module.

**Comment by kchibisov** [comment]
We actually could support it given that android works similar to the wayland layer-shell stuff. So I'd keep that open, since we might add support for that.

**Comment by kchibisov** [comment]
Similar as in it destroys window from time to time. The problem is that keyboard input will work strangely and mouse input as well.

I'll keep it open to until ether someone or I will try to add it and we'll see how it goes. Though it might be not added in the end given how weird it is.

**Comment by TheOnlyMrCat** [comment]
sctk 0.16 doesn't provide any abstraction for layer shells, but there is an implementation on master. If I understand correctly, this means this feature is waiting on the release of wayland-rs 0.30, as well as the refactor that update would require within winit code.

**Comment by kchibisov** [comment]
Yeah, I'm waiting for wayland-rs for a long time already.

**Comment by yavko** [comment]
> sctk 0.16 doesn't provide any abstraction for layer shells, but there is an implementation on master. If I understand correctly, this means this feature is waiting on the release of wayland-rs 0.30, as well as the refactor that update would require within winit code.

Once #2676 is merged, this won't be the case anymore, as sctk `0.17` is out, and it does provide the abstractions.

**Comment by Decodetalkers** [comment]
Now sctk 0.17 is merged.. Then what is going on? I cannot wait to use layershell

**Comment by yavko** [comment]
> Now sctk 0.17 is merged.. Then what is going on? I cannot wait to use layershell

No one is working on it, that's what's happening. If you want to use it, then add it yourself. Though if I have time I'll try to learn how winit works and *try* impl it myself

**Comment by RegenJacob** [comment]
Would be great to get [window::WindowLevel](https://docs.rs/winit/latest/winit/window/enum.WindowLevel.html#) on wayland

**Comment by ids1024** [comment]
I don't think that's possible, really. The `xdg_shell` protocol used on Wayland for normal windows doesn't support anything like `WindowLevel`. To do that you need layer-shell, but the behavior is different from a normal window so it wouldn't be as simple as having an option to set the level of a window.

**Comment by kchibisov** [comment]
No, the main issue is that you have no focus, your window gets randomly destroyed, etc, etc. Like this issue was discussed numerous times and was _postponed_ until 0.31.

**Comment by REALERvolker1** [comment]
It would be much MUCH better if it was implemented with missing features rather than waiting around until everything just works perfectly. Some of this stuff just won't work well in some compositors, but this protocol is essential for desktop programming with Iced and winit. Applications may not need these special extensions, but desktop components certainly do, and the only other viable alternative with layer-shell support is gtk3 (which I, for one, have several gripes with).

**Comment by REALERvolker1** [comment]
As for windows getting randomly destroyed and general instability, it would be much easier for the bugs to be resolved if more people used this feature. It would make us all more aware of bugs like these, and people can discuss and maybe fix them.

**Comment by Decodetalkers** [comment]
We have made layershell binding for iced, and make something like winit, in https://github.com/waycrate/exwlshelleventloop. We also bind session-lock. Winit should better just handle xdg window. Yes.. I agree, so we made this project, and made platform plugin for iced. Maybe you can take a try.

**Comment by LucaCoduriV** [comment]
@Decodetalkers OMG MAN Thank you !! I was looking for this for so long !
I would be happy to help you if needed.

**Comment by The-Briel-Deal** [comment]
I've set up layer-shell from the compositor side and client side using wlroots and wayland-rs using wayland-client which we use as a dependency for wayland in winit. 

Client implementation I made is in this repo https://github.com/The-Briel-Deal/gf-bar (which was just a small little status bar I made).

I haven't spent much time in the winit codebase but I might see if I can add an option in WindowAttributes to use layer shell as a POC here sometime this week.

I saw a few people saying that it wasn't possible because winit uses xdg-shell as it's role. That isn't a blocker at all though, you just need to request a different role during window creation and add the appropriate options in winit to set anchor, exclusive zone, margin, etc.

The one difficulty I see is if there is heavy usage of xdg-shell specific requests. But I'm hoping those can easily be disabled when you are using a layer-shell window.

**Comment by Ggcpp** [comment]
Any news on this?

**Comment by Lodjuret** [comment]
> Any news on this?

Seems to be worked on in #4044
